Only first chString in SimaticReadDataArea is updated in OpenHab

Hi,

I have a Thing with 2 string channels. But only the first string (with the lowest address) is updated.

```
channels:
  -id: ActualTime
    channelTypeUID: simatic:chString
    label: Actual time
    configuration:
      stateAddress: DB4.DBB2[4]
  -id: Sunrise
    channelTypeUID: simatic:chString
    label: Sun-rise
    configuration:
      stateAddress: DB4.DBB8[4]
``` 

When I checked the code I think there is a problem in the SimaticChannel.setState() method:

```java
public void setState(byte[] buffer, int start) {
        // logger.debug("item={}", toString());
        // logger.debug("buffer={}", buffer.length);
        // logger.debug("position={}", position);
        // logger.debug("item len={}", getStateAddress().getDataLength());

        int position = getStateAddress().getByteOffset() - start;

        try {
            ByteBuffer bb = ByteBuffer.wrap(buffer, position, getStateAddress().getDataLength());
            bb.order(ByteOrder.BIG_ENDIAN);

            if (isString()) {
                // check for '\0' char and resolve string length
                int i;
                for (i = position; i < getStateAddress().getDataLength(); i++) {
                    if (buffer[i] == 0) {
                        break;
                    }
                }
                String str = new String(buffer, position, i, thing.getCharset());
                setState(new StringType(str));
            } else if (isNumber()) {
```
For the string with the lowest address it works because
position = 2 - 2 => 0
Then the for loop will find the count of non-zero characters.
The new string will be: new String(buffer, 0, X, thing.getCharset()); //length X is somewhere between 0 and 3.

For the second string it does not work because
position = 8 - 2 => 6
**Then the for loop goes from i=6; i < 4**.
The new string will be: new String(buffer, 6, 6, thing.getCharset());

This will throw an exception (which I cant see in the production release of openhab, only when I debug it from eclipse. Don't know why):
(NOTE: This exception is with different addresses, but same situation)
```java
20:12:38.226 [ERROR] [atic.internal.simatic.SimaticChannel] - ChID=Sunset,StateAddress=DB4.DBB80[20],CmdAddress=null - Incoming data post error. Item:simatic:generic_device:PLC:PlcStatus:Sunset
java.lang.StringIndexOutOfBoundsException: Range [80, 80 + 80) out of bounds for length 100
	at jdk.internal.util.Preconditions$1.apply(Preconditions.java:55) ~[?:?]
	at jdk.internal.util.Preconditions$1.apply(Preconditions.java:52) ~[?:?]
	at jdk.internal.util.Preconditions$4.apply(Preconditions.java:213) ~[?:?]
	at jdk.internal.util.Preconditions$4.apply(Preconditions.java:210) ~[?:?]
	at jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:98) ~[?:?]
	at jdk.internal.util.Preconditions.outOfBoundsCheckFromIndexSize(Preconditions.java:118) ~[?:?]
	at jdk.internal.util.Preconditions.checkFromIndexSize(Preconditions.java:397) ~[?:?]
	at java.lang.String.checkBoundsOffCount(String.java:4853) ~[?:?]
	at java.lang.String.<init>(String.java:522) ~[?:?]
	at org.openhab.binding.simatic.internal.simatic.SimaticChannel.setState(SimaticChannel.java:393) ~[?:?]
	at org.openhab.binding.simatic.internal.simatic.SimaticTCP.readDataArea(SimaticTCP.java:335) ~[?:?]
	at org.openhab.binding.simatic.internal.simatic.SimaticGenericDevice.checkNewData(SimaticGenericDevice.java:402) ~[?:?]
	at org.openhab.binding.simatic.internal.simatic.SimaticGenericDevice.execute(SimaticGenericDevice.java:148) ~[?:?]
	at org.openhab.binding.simatic.internal.simatic.SimaticGenericDevice.lambda$0(SimaticGenericDevice.java:108) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[?:?]
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1583) [?:?]
``` 

Also, I think the condition for the loop should be **<=**
```java
                for (i = position; i <= getStateAddress().getDataLength(); i++) {
                    if (buffer[i] == 0) {
                        break;
                    }
                }
```
Otherwise it will always skip the last byte. But this is not important, and might be a breaking change for someone else using it.

Solution could be (I test it for my situation and it works):

```java
if (isString()) {
                // check for '\0' char and resolve string length
                int i;
                for (i = position; i < position + getStateAddress().getDataLength(); i++) {
                    if (buffer[i] == 0) {
                        break;
                    }
                }
                String str = new String(buffer, position, i - position, thing.getCharset());
                setState(new StringType(str));
```

--- a/org.openhab.binding.simatic/pom.xml
+++ b/org.openhab.binding.simatic/pom.xml
@@ -1,13 +1,13 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
-  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
 
   <modelVersion>4.0.0</modelVersion>
 
   <parent>
     <groupId>org.openhab.addons.bundles</groupId>
     <artifactId>org.openhab.addons.reactor.bundles</artifactId>
-      <version>4.0.1</version> 
+      <version>4.3.0</version> 
   </parent>
 
   <artifactId>org.openhab.binding.simatic</artifactId>

--- a/org.openhab.binding.simatic/src/main/history/dependencies.xml
+++ b/org.openhab.binding.simatic/src/main/history/dependencies.xml
@@ -1,9 +1,9 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
-<features xmlns="http://karaf.apache.org/xmlns/features/v1.6.0" name="org.openhab.binding.simatic-4.0.1">
+<features xmlns="http://karaf.apache.org/xmlns/features/v1.6.0" name="org.openhab.binding.simatic-4.3.0">
     <feature version="0.0.0">
         <feature>openhab-runtime-base</feature>
         <feature>wrap</feature>
-        <bundle>mvn:org.openhab.addons.bundles/org.openhab.binding.simatic/4.0.1</bundle>
-        <bundle>wrap:mvn:org.lastnpe.eea/eea-all/2.2.1</bundle>
+        <bundle>mvn:org.openhab.addons.bundles/org.openhab.binding.simatic/4.3.0</bundle>
+        <bundle>wrap:mvn:org.lastnpe.eea/eea-all/2.4.0</bundle>
     </feature>
 </features>

--- a/org.openhab.binding.simatic/src/main/java/org/openhab/binding/simatic/internal/simatic/SimaticChannel.java
+++ b/org.openhab.binding.simatic/src/main/java/org/openhab/binding/simatic/internal/simatic/SimaticChannel.java
@@ -385,8 +385,8 @@ public class SimaticChannel {
             if (isString()) {
                 // check for '\0' char and resolve string length
                 int i;
-                for (i = position; i < getStateAddress().getDataLength(); i++) {
-                    if (buffer[i] == 0) {
+                for (i = 0; i < getStateAddress().getDataLength(); i++) {
+                    if (buffer[position + i] == 0) {
                         break;
                     }
                 }

--- a/org.openhab.binding.simatic/src/main/java/org/openhab/binding/simatic/internal/simaticBindingConstants.java
+++ b/org.openhab.binding.simatic/src/main/java/org/openhab/binding/simatic/internal/simaticBindingConstants.java
@@ -25,7 +25,7 @@ import org.openhab.core.thing.type.ChannelTypeUID;
 @NonNullByDefault
 public class SimaticBindingConstants {
 
-    public static final String VERSION = "4.0.1";
+    public static final String VERSION = "4.3.0";
 
     private static final String BINDING_ID = "simatic";
 

